File: src/main/scala/poi/excel/GroupHelper.scala
package poi.excel

import poi.excel.api.ExcelReporter._

/**
 * Created with IntelliJ IDEA.
 * Author: shredinger
 * Date: 1/15/15
 * Time: 4:39 PM 
 * Project: scala-excel
 */
object GroupHelper {
  // Data array zipped with index
  type DataArrayZipIdx = List[(DataRow,Int)]

  /**
   * Describes a node in the heirarchy group
   * @param keyField key field id
   * @param keyVal key value
   * @param children  list of children nodes
   * @param leafs list of lear row ids (only when children.isEmpty)
   */
  case class GroupNode(keyField: String, keyVal:String, children:List[GroupNode], leafs: List[Int])

  case class MergeRegion(field:String, firstRow:Int, lastRow:Int)

  def createGroupNodes(reportInfo: ReportInfo, reportData: DataArray): List[GroupNode] = {
    // data fields zipped with index
    val dataFieldsZipIdx:DataArrayZipIdx = reportData.zipWithIndex

    // TBD: for a given row get a value by key
    //TODO: optimize
    lazy val dataFieldIndexById: Map[String, Int] = (reportInfo.dataFields.map(x => x.field) zipWithIndex).toMap
    def valueByField(row: DataRow, id:String) : String = {
      row(dataFieldIndexById.get(id).get)
    }

    val groupFields = reportInfo.rowGroup.map(f => f.field)

    val groupsNodes:List[GroupNode] = groupFields.map { fld =>
      // TODO must be implemeneted
      null
    }

    def createGroupNodes(dataArray:DataArrayZipIdx, groupFields:List[String]):List[GroupNode] = {

      dataArray.groupBy(record => valueByField(record._1, groupFields.head))
        .map(x => GroupNode(groupFields.head, x._1,
          groupFields.tail match {
            case Nil => List()
            case _ => createGroupNodes(x._2, groupFields.tail)
          },
          groupFields.tail match {
            case Nil => x._2.map(dataRowWithIndex => dataRowWithIndex._2)
            case _ => Nil
          })
        )
        .toList.sortWith((l, r) => l.keyVal < r.keyVal)
    }

    createGroupNodes(dataFieldsZipIdx, groupFields)
  }

  def getMergeRegions(groupNodes: List[GroupNode]):List[MergeRegion] = {
    groupNodes.map(groupNode =>
      groupNode.leafs match {
        case Nil => {
          val leaves = getAllLeaves(groupNode.children)
          MergeRegion(groupNode.keyField, leaves.min, leaves.max) :: getMergeRegions(groupNode.children)
        }
        case _ => MergeRegion(groupNode.keyField, groupNode.leafs.min, groupNode.leafs.max) :: Nil
      }).flatten
  }

  private def getAllLeaves(children: List[GroupNode]): List[Int] = {
    children.map(node => node.leafs match {
      case Nil => getAllLeaves(node.children)
      case _ => node.leafs
    }).flatten.sorted
  }
}

File: src/main/java/poi/excel/report/GroupHelper.java
package poi.excel.report;

import com.google.common.collect.ArrayListMultimap;
import poi.excel.report.api.DataField;
import poi.excel.report.api.ReportField;
import poi.excel.report.api.ReportInfo;
import poi.excel.report.util.Mapper;
import poi.excel.report.util.Utils;
import poi.excel.report.util.Utils.Tuple;

import java.util.*;

/**
 * Created with IntelliJ IDEA.
 * Author: shredinger
 * Date: 1/18/15
 * Time: 9:14 PM
 * Project: scala-excel
 */
public class GroupHelper {

    public static List<GroupNode> createGroupNodes(ReportInfo reportInfo, List<List<String>> reportData){


        Map<String, Integer> dataFieldIndexById = new HashMap<>();
        List<DataField> dataFields = reportInfo.getDataFields();
        for(int i = 0; i < dataFields.size(); i ++){
            dataFieldIndexById.put(dataFields.get(i).getField(), i);
        }

        List<String> groupFields = Utils.map(reportInfo.getRowGroup(), new Mapper<ReportField, String>() {
            @Override
            public String apply(ReportField value) {
                return value.getField();
            }
        });

        List<Utils.Tuple<List<String>, Integer>> rowsWithIndex = zipWithIndex(reportData);

        return createGroupNodes(rowsWithIndex, dataFieldIndexById, groupFields);
    }

    public static List<MergeRegion> getMergeRegions(List<GroupNode> groupNodes){
        ArrayList<MergeRegion> mergeRegions = new ArrayList<>();
        for(GroupNode groupNode: groupNodes){
            List<Integer> leaves = groupNode.getLeaves();
            if(leaves.isEmpty()){
                List<Integer> allLeaves = getAllLeaves(groupNode.getChildren());
                int minRow = Collections.min(allLeaves);
                int maxRow = Collections.max(allLeaves);
                mergeRegions.add(new MergeRegion(groupNode.getKey(), minRow, maxRow));
                mergeRegions.addAll(getMergeRegions(groupNode.getChildren()));
            } else {
                int firstRow = Collections.min(leaves);
                int lastRow = Collections.max(leaves);
                mergeRegions.add(new MergeRegion(groupNode.getKey(), firstRow, lastRow));
            }
        }

        return mergeRegions;
    }

    private static List<Integer> getAllLeaves(List<GroupNode> children){
        ArrayList<Integer> leaves = new ArrayList<>();
        for(GroupNode groupNode: children){
            if(groupNode.getLeaves().isEmpty()){
                leaves.addAll(getAllLeaves(groupNode.getChildren()));
            }else{
                leaves.addAll(groupNode.getLeaves());
            }
        }
        return leaves;
    }

    private static List<GroupNode> createGroupNodes(Collection<Tuple<List<String>, Integer>> dataArray,
                                                    Map<String, Integer> dataFieldsByIndex,
                                                    List<String> groupFields){
        ArrayList<GroupNode> groupNodes = new ArrayList<>();
        String groupField = groupFields.get(0);
        Map<String, Collection<Tuple<List<String>, Integer>>> groupedDataArray = group(dataArray, dataFieldsByIndex, groupField);
        for(String value: groupedDataArray.keySet()){
            Collection<Tuple<List<String>, Integer>> rowsForKey = groupedDataArray.get(value);


            List<Integer> leaves = groupFields.size() == 1 ? getRowIndexes(rowsForKey) : new ArrayList<Integer>();
            List<GroupNode> children = groupFields.size() == 1 ?
                    new ArrayList<GroupNode>() :
                    createGroupNodes(rowsForKey, dataFieldsByIndex, groupFields.subList(1, groupFields.size()));
            groupNodes.add(new GroupNode(groupField, value, children, leaves));
        }

        sortGroupNodes(groupNodes);

        return groupNodes;
    }

    private static List<GroupNode> sortGroupNodes(List<GroupNode> groupNodes){
        Collections.sort(groupNodes, new Comparator<GroupNode>() {
            @Override
            public int compare(GroupNode o1, GroupNode o2) {
                return o1.getValue().compareTo(o2.getValue());
            }
        });
        return groupNodes;
    }

    private static List<Integer> getRowIndexes(Collection<Tuple<List<String>, Integer>> rows){
        ArrayList<Integer> rowIndexes = new ArrayList<>();
        for (Tuple<List<String>, Integer> rowWithIndexes: rows){
            rowIndexes.add(rowWithIndexes.getFiled2());
        }
        return rowIndexes;
    }

    private static Map<String, Collection<Tuple<List<String>, Integer>>> group(Collection<Tuple<List<String>, Integer>> dataArray,
                                                                                      Map<String, Integer> dataFieldIndexById,
                                                                                      String groupField){
        ArrayListMultimap<String, Tuple<List<String>, Integer>> groupedRows = ArrayListMultimap.create();
        for(Tuple<List<String>, Integer> rowWithIndex: dataArray){
            Integer indexInRow = dataFieldIndexById.get(groupField);
            String cellValue = rowWithIndex.getFiled1().get(indexInRow);
            groupedRows.put(cellValue, rowWithIndex);
        }
        return groupedRows.asMap();

    }

    private static List<Utils.Tuple<List<String>, Integer>> zipWithIndex(List<List<String>> dataArray){
        ArrayList<Utils.Tuple<List<String>, Integer>> rowsWithIndex = new ArrayList<>();
        for(int row = 0; row < dataArray.size(); row ++){
            rowsWithIndex.add(new Utils.Tuple<List<String>, Integer>(dataArray.get(row), row));
        }
        return rowsWithIndex;
    }


}
